core/src/protocols_handler/mod.rs

```diff
@@ -108,22 +108,35 @@ pub trait ProtocolsHandler {
     /// and will be passed back in `inject_substream` or `inject_outbound_closed`.
     type OutboundOpenInfo;
 
-    /// Produces a `ConnectionUpgrade` for the protocol or protocols to accept when listening.
+    /// The timeout for protocol negotiation on inbound substreams.
+    fn inbound_timeout(&self) -> Duration {
+        Duration::from_secs(10)
```

Comment:
What about changing `listen_protocol()` to return a tuple of `(Self::InboundProtocol, Duration)` instead?

Reply:
Thanks for the suggestion, I haven't though of that. I can try going down that route, though it does mean a breaking API change for `ProtocolsHandler`, if we're ok with that.

Reply:
Breaking changes are ok as long as they are not too large and we properly document in the CHANGELOG. Also, we're already breaking the `Transport` trait in this release.